File: OfflineTranscription/Utilities/WAVWriter.swift
import Foundation

enum WAVWriter {
    /// Write 16kHz mono 16-bit PCM WAV file from Float32 samples normalized to [-1, 1].
    static func write(samples: [Float], sampleRate: Int = 16000, to url: URL) throws {
        let numChannels: UInt16 = 1
        let bitsPerSample: UInt16 = 16
        let byteRate = UInt32(sampleRate) * UInt32(numChannels) * UInt32(bitsPerSample / 8)
        let blockAlign = numChannels * (bitsPerSample / 8)
        let dataSize = UInt32(samples.count * Int(blockAlign))
        let fileSize = 36 + dataSize

        var data = Data()
        data.reserveCapacity(Int(44 + dataSize))

        // RIFF header
        data.append(contentsOf: "RIFF".utf8)
        data.appendLittleEndian(fileSize)
        data.append(contentsOf: "WAVE".utf8)

        // fmt sub-chunk
        data.append(contentsOf: "fmt ".utf8)
        data.appendLittleEndian(UInt32(16))
        data.appendLittleEndian(UInt16(1)) // PCM
        data.appendLittleEndian(numChannels)
        data.appendLittleEndian(UInt32(sampleRate))
        data.appendLittleEndian(byteRate)
        data.appendLittleEndian(blockAlign)
        data.appendLittleEndian(bitsPerSample)

        // data sub-chunk
        data.append(contentsOf: "data".utf8)
        data.appendLittleEndian(dataSize)

        // Convert Float32 → Int16 PCM
        for sample in samples {
            let clamped = max(-1.0, min(1.0, sample))
            let int16Value = Int16(clamped * 32767.0)
            data.appendLittleEndian(int16Value)
        }

        try data.write(to: url, options: .atomic)
    }
}

private extension Data {
    mutating func appendLittleEndian<T: FixedWidthInteger>(_ value: T) {
        var le = value.littleEndian
        withUnsafePointer(to: &le) { ptr in
            append(UnsafeBufferPointer(start: ptr, count: 1))
        }
    }
}

File: VoicePingIOSAndroidOfflineSpeechTranslationAndroid/app/src/main/kotlin/com/jima/offlinetranscription/util/WavWriter.kt
package com.voiceping.offlinetranscription.util

import com.voiceping.offlinetranscription.service.AudioConstants
import java.io.BufferedOutputStream
import java.io.File
import java.io.FileOutputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * Writes Float32 audio samples to a 16-bit PCM WAV file (16kHz mono).
 */
object WavWriter {

    private const val HEADER_SIZE = 44
    private const val BITS_PER_SAMPLE = 16
    private const val NUM_CHANNELS = 1
    private const val CHUNK_SIZE = 4096

    fun write(samples: FloatArray, sampleRate: Int = AudioConstants.SAMPLE_RATE, outputFile: File) {
        val dataSize = samples.size * (BITS_PER_SAMPLE / 8) * NUM_CHANNELS
        val fileSize = HEADER_SIZE + dataSize

        outputFile.parentFile?.mkdirs()

        BufferedOutputStream(FileOutputStream(outputFile)).use { out ->
            // Write WAV header
            out.write(buildHeader(sampleRate, dataSize, fileSize))

            // Write PCM data in chunks to avoid doubling memory
            val buf = ByteBuffer.allocate(CHUNK_SIZE * 2).order(ByteOrder.LITTLE_ENDIAN)
            var offset = 0
            while (offset < samples.size) {
                buf.clear()
                val end = minOf(offset + CHUNK_SIZE, samples.size)
                for (i in offset until end) {
                    val clamped = samples[i].coerceIn(-1f, 1f)
                    val int16 = (clamped * 32767f).toInt().toShort()
                    buf.putShort(int16)
                }
                out.write(buf.array(), 0, buf.position())
                offset = end
            }
        }
    }

    private fun buildHeader(sampleRate: Int, dataSize: Int, fileSize: Int): ByteArray {
        val byteRate = sampleRate * NUM_CHANNELS * (BITS_PER_SAMPLE / 8)
        val blockAlign = NUM_CHANNELS * (BITS_PER_SAMPLE / 8)

        return ByteBuffer.allocate(HEADER_SIZE).order(ByteOrder.LITTLE_ENDIAN).apply {
            // RIFF chunk
            put("RIFF".toByteArray(Charsets.US_ASCII))
            putInt(fileSize - 8)
            put("WAVE".toByteArray(Charsets.US_ASCII))

            // fmt sub-chunk
            put("fmt ".toByteArray(Charsets.US_ASCII))
            putInt(16)           // sub-chunk size
            putShort(1)          // PCM format
            putShort(NUM_CHANNELS.toShort())
            putInt(sampleRate)
            putInt(byteRate)
            putShort(blockAlign.toShort())
            putShort(BITS_PER_SAMPLE.toShort())

            // data sub-chunk
            put("data".toByteArray(Charsets.US_ASCII))
            putInt(dataSize)
        }.array()
    }
}
